File: 0045-jump-game-ii/0045-jump-game-ii.cpp
class Solution {
public:
    int dp[10001];
    int n;

    int solve(int idx, vector<int>& nums) {
        if (idx == n - 1) return 0;

        if (dp[idx] != -1) return dp[idx];

        int ans = INT_MAX;

        for (int i = idx+1; i<=min(n-1, idx+nums[idx]); i++) {
            int next = solve(i,nums);
            if (next != INT_MAX) ans = min(ans,1+next);
        }

        return dp[idx] = ans;
    }

    int jump(vector<int>& nums) {
        n = nums.size();
        memset(dp, -1, sizeof(dp));
        return solve(0, nums);
    }
};

File: 0045-jump-game-ii/0045-jump-game-ii.kt
class Solution {
    
    fun jump(nums: IntArray): Int {
        val n = nums.size
        val dp = IntArray(n) { -1 }

        fun solve(idx: Int): Int {
            if (idx == n - 1) return 0

            if (dp[idx] != -1) return dp[idx]

            var ans = Int.MAX_VALUE

            for (i in idx + 1..minOf(n - 1, idx + nums[idx])) {
                val next = solve(i)
                if (next != Int.MAX_VALUE) {
                    ans = minOf(ans, 1 + next)
                }
            }

            dp[idx] = ans
            return ans
        }

        return solve(0)
    }
}